File: data_structures/linked_list/ruby/linked_list.rb
class Node
    attr_writer :next
    attr_reader :next, :data

    def initialize(data, next_node = nil)
        @data = data
        @next = next_node
    end
end

class LinkedList
    def initialize
        @head = nil
        @tail = @head
    end

    def insertFront(data)
        node = Node.new(data, @head)
        @tail = node if @tail == nil
        @head = node
    end

    def insertBack(data)
        node = Node.new(data)
        @head = node if @head == nil
        @tail.next = node if @tail != nil
        @tail = node
    end

    def find(data)
        current = @head
        while current != nil do
            return current if data == current.data
            current = current.next
        end
        return nil
    end

    def remove(node)
        prev = nil
        current = @head

        while current != nil do
            if node.equal? current
                if current == @head
                    @head = @head.next
                elsif current == @tail
                    @tail = prev
                    prev.next = nil
                else
                    prev.next = current.next
                end
                break
            end
            prev = current
            current = current.next
        end
    end

    def iterate
        current = @head
        while current != nil do
            yield current.data
            current = current.next
        end
    end
end

# Usage
L = LinkedList.new
L.insertBack(3)
L.insertBack(4)
L.insertFront(2)
L.insertFront(1)

# 1 2 3 4
L.iterate { |data| puts(data) }

node_2 = L.find(4)
L.remove(node_2)

# 1 3 4
L.iterate { |data| puts(data) }

File: data_structures/linked_list/python/linked_list.py
class Node:
    def __init__(self, data=None, next_node=None):
        self.data = data
        self.next = next_node
        self.prev = None

    def __repr__(self):
        return self.data or ''


class LinkedList:
    '''
    Implementation of Linked list data structure
    '''

    def __init__(self, data_iterable):
        self.head = None
        for data in data_iterable:
            self.add(data)

    def add(self, data):
        node = Node(data)
        node.next = self.head
        self.head = node
        if node.next:
            node.next.prev = node

    def search(self, data):
        node = self.head
        while node:
            if node.data == data:
                return node
            node = node.next
        return None

    def remove(self, node):
        if node is self.head:
            self.head = node.next
        else:
            node.prev.next = node.next
        if node.next:
            node.next.prev = node.prev
    
    def __repr__(self):
        node = self.head
        data = []
        while node:
            data.append(node.data)
            node = node.next
        return ' '.join(data)


if __name__ == '__main__':
    l = LinkedList(['d', 'c', 'b'])
    assert str(l) == 'b c d'
    l.add('a')
    assert str(l) == 'a b c d'
    l.remove(l.search('a'))
    l.remove(l.search('c'))
    assert str(l) == 'b d'

